teuthology/openstack/__init__.py

```diff
@@ -128,6 +136,42 @@ def get_addresses(self):
                     return self['addresses']
                 self.set_info()
 
+    @staticmethod
+    def parse_openstack(data):
+        """
+        Parse OpenStack structures, like multiline data:
+           a='10', b='11', c='13'
+           a='20', b='21', c='23'
+        """
+        values=[{a:b.strip("'") for a,b in
+                    [p.split("=") for p in pairs.split(", ")]}
+                                    for pairs in data.split("\n")]
+        return values
+
+    def get_ip_openstack(self):
+        subnets = json.loads(misc.sh("openstack subnet list -f json --ip-version 4"))
+        subnet_ids = [i['ID'] for i in subnets]
+
+        if not subnet_ids:
+            raise Exception("no subnet with ip_version == 4")
+
+        ports = json.loads(misc.sh("openstack port list -f json -c fixed_ips -c device_id"))
+        fixed_ips = None
+        for port in ports:
+            if port['device_id'] == self['id']:
+                fixed_ips = self.parse_openstack(port['Fixed IP Addresses'])
+                break
+        if not fixed_ips:
+            raise Exception("no fixed ip record found")
+        ip = None
+        for record in fixed_ips:
+            if record['subnet_id'] in subnet_ids:
+                ip = record['ip_address']
+                break
+        if not ip:
+            raise Exception("no ip")
+        return ip
+
     def get_ip_neutron(self):
```

Comment:
Since `python-neutron` is being dropped as a dependency by this PR, would it be a good idea to mention that here? I.e. I'm assuming this function will no longer work without the dependency.

Reply:
That makes sense, but I also like to drop this method in separate commit, in case there are some changes upstream so we can ease this patch merge upstream.